src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -274,11 +278,30 @@ public CompiledScript compileInternal(Script script) {
         CompiledScript compiled = cache.getIfPresent(cacheKey);
         if (compiled == null) {
             //Either an un-cached inline script or an indexed script
-            compiled = new CompiledScript(lang, scriptEngineService.compile(code));
+            try {
+                compiled = new CompiledScript(script.getType(), name, lang, scriptEngineService.compile(code));
+            } catch (Exception e) {
+                String message = "Compilation error with " + script.getType() + " script ";
+                if (name != null) {
+                    message += "[" + name + "] ";
+                }
+                message += "using [" + lang + "]: " + ExceptionsHelper.detailedMessage(e);
+                throw new ScriptException(message, e);
+            }
             //Since the cache key is the script content itself we don't need to
             //invalidate/check the cache if an indexed script changes.
             cache.put(cacheKey, compiled);
         }
+
+        // An indexed or inline script may be mistaken for the other since the cache key relies on
+        // only the code of the script. Two indexed scripts may also have the same code, but different
+        // names. If this is case, the name and type parameters must be modified accordingly.
+        // Then cache the most recent version used.
```

Comment:
@rjernst I'll give this some thought.  Maybe it makes sense to turn cache key into a method that just returns a concatenated string instead of a full blown object that isn't really necessary?